torch/csrc/jit/profiling_graph_executor_impl.cpp

```diff
@@ -8,22 +11,34 @@ bool& getProfilingMode() {
   return profiling_mode;
 }
 
+static std::shared_ptr<Graph> prepareGraph(const std::shared_ptr<Graph>& graph) {
+  auto g = graph->copy();
+  runRequiredPasses(g);
+  return g;
+}
+
+ProfilingGraphExecutorImpl::ProfilingGraphExecutorImpl(const std::shared_ptr<Graph>& graph, bool optimize)
+: GraphExecutorImplBase(graph, optimize),
+  pr_(ProfilingRecord::instrumentGraph(prepareGraph(graph))),
+  profiling_plan_(pr_->graph()) {}
+
 ExecutionPlan ProfilingGraphExecutorImpl::getPlanFor(Stack& stack) {
-  {
-    std::lock_guard<std::mutex> lock(compile_mutex);
-    if (!pr_) {
-      auto g = graph->copy();
-      runRequiredPasses(g);
-      pr_ = ProfilingRecord::instrumentGraph(g);
-      exec_plan_ = caffe2::make_unique<ExecutionPlan>(pr_->profiled_graph_);
-    }
+  if (optimized_plan_) {
+    return *optimized_plan_;
   }
-  if (pr_->profiling_count_ > 0) {
-    return *exec_plan_;
+  if (!pr_->ready()) {
+    return profiling_plan_;
   }
-  AT_ERROR("Not yet implemented");
+
+  auto copy = pr_->graph()->copy();
+  InsertGuards(copy);
+  EliminateGuards(copy);
```

Comment:
Nit: maybe it should be called `EliminateRedundantGuards` or something like that. Otherwise, it seems that these two lines should cancel each other:)

Reply:
renamed it